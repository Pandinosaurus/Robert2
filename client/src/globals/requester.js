import axios from 'axios';
import deepObjectSet from 'lodash/set';
import isPlainObject from 'lodash/isPlainObject';
import flattenObject from '@/utils/flattenObject';
import config from '@/globals/config';
import cookies from '@/utils/cookies';
import DateTime from '@/utils/datetime';
import Day from '@/utils/day';

const requester = axios.create({
    baseURL: config.api.url,
    headers: config.api.headers,
});

requester.interceptors.request.use(
    (_request) => {
        const { params, onProgress, ...request } = _request;

        // - Ajoute une méthode haut-niveau permettant de récupérer l'avancement sous forme numérique.
        //   - Si la méthode de la requête est POST, PUT ou PATCH, c'est l'événement `onUploadProgress` qui sera écouté.
        //   - Sinon ce sera `onDownloadProgress`
        if (onProgress) {
            const progressCallback = (event) => {
                if (!event.lengthComputable) {
                    return;
                }

                const { loaded, total } = event;
                const percent = (loaded / total) * 100;
                onProgress(percent);
            };

            if (request.method && ['POST', 'PUT', 'PATCH'].includes(request.method.toUpperCase())) {
                request.onUploadProgress = progressCallback;
            } else {
                request.onDownloadProgress = progressCallback;
            }
        }

        // - Traitement spécial des objets littéraux contenant des fichiers.
        //   => On utilise un envoi via `multipart/form-data` et on met les données
        //      supplémentaires éventuelles dans une clé spéciale `@data`)
        const hasFiles = (
            isPlainObject(request.data) &&
            Object.values(request.data).some((value) => (
                value instanceof File
            ))
        );
        if (hasFiles) {
            const formData = new FormData();
            const data = Object.entries(flattenObject(request.data)).reduce(
                (_data, [key, value]) => {
                    if (value instanceof File) {
                        formData.append(btoa(key), value);
                        return _data;
                    }
                    return deepObjectSet(_data, key, value);
                },
                {},
            );

            formData.append('@data', JSON.stringify(data));
            request.data = formData;
        }

        if (params) {
            Object.keys(params).forEach((name) => {
                if (params[name] === true) {
                    params[name] = '1';
                }
                if (params[name] === false) {
                    params[name] = '0';
                }
                if (params[name] instanceof DateTime || params[name] instanceof Day) {
                    params[name] = params[name].toString();
                }
            });
            request.params = params;
        }

        const authToken = cookies.get(config.auth.cookie);
        if (authToken) {
            request.headers.Authorization = `Bearer ${authToken}`;
        }
        return request;
    },
    (error) => Promise.reject(error),
);

export default requester;
