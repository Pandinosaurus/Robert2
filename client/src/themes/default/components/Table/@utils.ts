import type { ColumnsVisibility } from 'vue-tables-2-premium';

export type ColumnsDisplay = Record<string, boolean>;

const STORAGE_KEY_PREFIX = 'vuetables_';

const initFromStorage = (tableName: string, columns: ColumnsDisplay): ColumnsDisplay => {
    const storedTableState = localStorage.getItem(`${STORAGE_KEY_PREFIX}${tableName}`);
    if (!storedTableState) {
        return columns;
    }

    let tableState;
    try {
        tableState = JSON.parse(storedTableState);
    } catch {
        localStorage.removeItem(`${STORAGE_KEY_PREFIX}${tableName}`);
        return columns;
    }

    if (!tableState) {
        return columns;
    }

    const { userColumnsDisplay } = tableState;
    if (!userColumnsDisplay) {
        return columns;
    }

    return Object.fromEntries(
        Object.keys(columns).map((column: string) => (
            [column, userColumnsDisplay.includes(column)]
        )),
    );
};

export const initColumnsDisplay = (tableName: string | undefined, columns: ColumnsDisplay): ColumnsVisibility => (
    Object.fromEntries(
        Object
            .entries((
                tableName !== undefined
                    ? initFromStorage(tableName, columns)
                    : columns
            ))
            .map(([name, visible]: [string, boolean]) => (
                [name, `${visible ? 'min' : 'max'}_mobileP`]
            )),
    )
);
