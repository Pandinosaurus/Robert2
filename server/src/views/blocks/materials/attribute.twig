{#
    À cause de la mauvaise prise en charge du `white-space: nowrap` par DomPDF,
    les espaces invisibles doivent être supprimés pour éviter qu'ils soient wrappés
    n'importe comment après le label et avant la valeur, mais qu'il puissent être
    wrappés dans la valeur elle-même si c'est un texte un peu long.
#}
<span class="lighter">
    {{- attribute['name'] }}&nbsp;:&nbsp;
    {%- if attribute['type'] == 'boolean' -%}
    {{- attribute['value'] == '1' ? __('yes') : __('no') -}}.
    {%- elseif attribute['type'] == 'date' -%}
    {{- attribute['value']|format_date('short', locale=locale) -}}.
    {%- else -%}
    {{- attribute['value'] -}}
    {%- endif -%}
    {%- if attribute['unit'] is not empty %}&nbsp;{{ attribute['unit'] }}.{% endif -%}
</span>
