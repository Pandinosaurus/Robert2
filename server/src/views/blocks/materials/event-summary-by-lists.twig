{% for listName, listMaterials in materialsSorted -%}
{% if materialsSorted|length > 1 -%}
<h2>
    {{-
        not listName
            ? __('materials-list-global')
            : __('materials-list-name', listName)
    -}}
</h2>
{% else -%}
<h2>{{ __('materials') }}</h2>
{% endif -%}
{%
    set colspan = 4
        + (showPictures ? 1 : 0)
        + ((showReplacementPrices ? 1 : 0) * 2)
        + (showDescriptions ? 1 : 0)
        + (showTags ? 1 : 0)
%}
<table class="listing-table">
    <thead>
        <tr>
            {% if showPictures -%}
            <th>{{ __('picture') }}</th>
            {% endif -%}
            <th>{{ __('ref') }}</th>
            <th>{{ __('designation') }}</th>
            {% if showDescriptions -%}
            <th>{{ __('description') }}</th>
            {% endif -%}
            {% if showTags -%}
            <th class="center">{{ __('tags') }}</th>
            {% endif -%}
            {% if showReplacementPrices -%}
            <th class="right">{{ __('replacement-value') }}</th>
            <th class="right">{{ __('total-replacement-value') }}</th>
            {% endif -%}
            <th class="center">{{ __('qty') }}</th>
            <th class="center">{{ __('qty-returned') }}</th>
        </tr>
    </thead>
    <tbody>
        {% if materialDisplayMode == 'categories' %}
            {%- for category, materials in listMaterials -%}
                <tr>
                    <td class="section-title" colspan="{{ colspan }}">
                        {{- category|replace({ '__other': __('not-categorized') }) -}}
                    </td>
                </tr>
                {% for material in materials %}
                    {{- include('blocks/materials/line-event-summary.twig', { material }) -}}
                {% endfor %}
            {%- endfor -%}
        {%- elseif materialDisplayMode == 'sub-categories' -%}
            {%- for subCategory, materials in listMaterials -%}
                <tr>
                    <td class="section-title" colspan="{{ colspan }}">
                        {%- if subCategory == '--' -%}
                            {{- __('not-categorized') -}}
                        {%- else -%}
                            {{- subCategory|replace({ '__other': __('other-materials') }) -}}
                        {%- endif -%}
                    </td>
                </tr>
                {% for material in materials %}
                    {{- include('blocks/materials/line-event-summary.twig', { material }) -}}
                {% endfor %}
            {%- endfor -%}
        {%- elseif materialDisplayMode == 'parks' -%}
            {%- for park, materials in listMaterials -%}
                <tr>
                    <td class="section-title" colspan="{{ colspan }}">
                        {%- if park is not empty -%}
                            {{- park -}}
                        {%- else -%}
                            ({{ __('unknown-park') }})
                        {%- endif -%}
                    </td>
                </tr>
                {% for material in materials %}
                    {{-
                        include('blocks/materials/line-event-summary.twig', {
                            withoutPark: true,
                            material,
                        })
                    -}}
                {% endfor %}
            {%- endfor -%}
        {%- else -%}
            {% for material in listMaterials %}
                {{- include('blocks/materials/line-event-summary.twig', { material }) -}}
            {% endfor %}
        {%- endif %}
    </tbody>
</table>
{% endfor %}