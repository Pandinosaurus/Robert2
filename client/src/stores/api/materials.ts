import requester from '@/globals/requester';
import { z } from '@/utils/validation';
import { omit } from 'lodash';
import { TagSchema } from './tags';
import { DocumentSchema } from './documents';
import { AttributeWithValueSchema } from './attributes';
import { BookingSummarySchema, createBookingSummarySchema } from './bookings';
import { withPaginationEnvelope } from './@schema';

import type Period from '@/utils/period';
import type { ProgressCallback, AxiosRequestConfig as RequestConfig } from 'axios';
import type { PaginatedData, SortableParams, PaginationParams } from './@types';
import type { Park } from './parks';
import type { Category } from './categories';
import type { Event } from './events';
import type { SubCategory } from './subcategories';
import type { Attribute, AttributeWithValue } from './attributes';
import type { Tag } from './tags';
import type { Document } from './documents';
import type { SchemaInfer } from '@/utils/validation';
import type { ZodRawShape } from 'zod';
import type { Simplify } from 'type-fest';
import type { DegressiveRate } from './degressive-rates';
import type { Tax } from './taxes';

// ------------------------------------------------------
// -
// -    Schema / Enums
// -
// ------------------------------------------------------

/** Représente le matériel non catégorisés. */
export const UNCATEGORIZED = 'uncategorized';

//
// - Schemas secondaires
//

const MaterialBaseSchema = z.strictObject({
    id: z.number(),
    name: z.string(),
    reference: z.string(),
    park_id: z.number(),
    picture: z.string().nullable(),
    description: z.string().nullable(),
    category_id: z.number().nullable(),
    sub_category_id: z.number().nullable(),
    rental_price: z.decimal().nullable().optional(),
    degressive_rate_id: z.number().nullable().optional(),
    tax_id: z.number().nullable().optional(),
    replacement_price: z.decimal().nullable(),
    stock_quantity: z.number().nullable().transform(
        (value: number | null): number => value ?? 0,
    ),
    out_of_order_quantity: z.number().nullable().transform(
        (value: number | null): number => value ?? 0,
    ),
    is_hidden_on_bill: z.boolean().optional(),
    is_discountable: z.boolean().optional(),
    attributes: z.lazy(() => AttributeWithValueSchema.array()),
    tags: z.lazy(() => TagSchema.array()),
    note: z.string().nullable(),
    created_at: z.datetime(),
    updated_at: z.datetime().nullable(),
});

const createMaterialSchemaFactory = <T extends ZodRawShape>(augmentation: T) => (
    <InnerT extends ZodRawShape>(innerAugmentation: InnerT) => (
        MaterialBaseSchema
            .extend<T>(augmentation)
            .extend<InnerT>(innerAugmentation)
    )
);

export const createMaterialSchema = createMaterialSchemaFactory({});

export const createMaterialDetailsSchema = createMaterialSchemaFactory(
    {
        available_quantity: z.number(),
        departure_inventory_todo: z.lazy(() => BookingSummarySchema).nullable(),
        return_inventory_todo: z.lazy(() => BookingSummarySchema).nullable(),
    },
);

export const createMaterialWithAvailabilitySchema = createMaterialSchemaFactory(
    {
        available_quantity: z.number(),
        is_deleted: z.boolean(),
    },
);

export const createMaterialWithContextExcerptSchema = createMaterialSchemaFactory(
    {
        degressive_rate: z.decimal().nullable().optional(),
        rental_price_period: z.decimal().nullable().optional(),
        is_deleted: z.boolean(),
    },
);

export const createMaterialWithContextSchema = createMaterialSchemaFactory(
    {
        degressive_rate: z.decimal().nullable().optional(),
        rental_price_period: z.decimal().nullable().optional(),
        available_quantity: z.number(),
        is_deleted: z.boolean(),
    },
);

const MaterialBookingSummarySchema = z.lazy(() => (
    createBookingSummarySchema({
        pivot: z.strictObject({
            quantity: z.number().positive(),
        }),
    })
));

//
// - Schemas principaux
//

export const MaterialSchema = createMaterialSchema({});

export const MaterialDetailsSchema = createMaterialDetailsSchema({});

export const MaterialWithAvailabilitySchema = createMaterialWithAvailabilitySchema({});
export const MaterialWithContextExcerptSchema = createMaterialWithContextExcerptSchema({});
export const MaterialWithContextSchema = createMaterialWithContextSchema({});

export const MaterialPublicSchema = (() => {
    const baseSchema = MaterialBaseSchema.extend({
        degressive_rate: z.decimal(),
        rental_price_period: z.decimal().nullable(),
        available_quantity: z.number(),
    });

    return baseSchema.pick({
        id: true,
        name: true,
        description: true,
        picture: true,
        degressive_rate: true,
        available_quantity: true,
        rental_price: true,
        rental_price_period: true,
    });
})();

// ------------------------------------------------------
// -
// -    Types
// -
// ------------------------------------------------------

export type Material = SchemaInfer<typeof MaterialSchema>;

export type MaterialDetails = SchemaInfer<typeof MaterialDetailsSchema>;

export type MaterialWithAvailability = SchemaInfer<typeof MaterialWithAvailabilitySchema>;

export type MaterialWithContext = SchemaInfer<typeof MaterialWithContextSchema>;

export type MaterialPublic = SchemaInfer<typeof MaterialPublicSchema>;

//
// - Secondary types.
//

export type MaterialBookingSummary = SchemaInfer<typeof MaterialBookingSummarySchema>;

//
// - Edition
//

type MaterialEditAttribute = {
    id: Attribute['id'],
    value: AttributeWithValue['value'],
};

export type MaterialEdit = {
    name: string | null,
    picture?: File | null,
    reference: string | null,
    description: string | null,
    park_id: Park['id'] | null,
    category_id: Category['id'] | null,
    sub_category_id: SubCategory['id'] | null,
    rental_price: string | null,
    degressive_rate_id: DegressiveRate['id'] | null,
    tax_id: Tax['id'] | null,
    stock_quantity: string | number | null,
    out_of_order_quantity: string | number | null,
    replacement_price: string | null,
    is_hidden_on_bill: boolean,
    is_discountable: boolean,
    note: string | null,
    tags?: Array<Tag['id']>,
    attributes?: MaterialEditAttribute[],
};

//
// - Récupération
//

export type BaseFilters = Nullable<{
    search?: string,
    category?: Category['id'],
    subCategory?: SubCategory['id'],
}>;

export type Filters = Simplify<(
    & Omit<BaseFilters, 'category'>
    & Nullable<{
        quantitiesPeriod?: Period,
        category?: Category['id'] | typeof UNCATEGORIZED,
        park?: Park['id'],
        tags?: Array<Tag['id']>,
    }>
)>;

type GetAllParamsBase = Filters & SortableParams & { withDeleted?: boolean, onlyDeleted?: boolean };
type GetAllParamsPaginated = GetAllParamsBase & PaginationParams & { paginated?: true };
type GetAllParamsRaw = GetAllParamsBase & { paginated: false };

type GetBookingsParams = (
    & PaginationParams
    & { period?: Period }
    & Pick<SortableParams, 'ascending'>
);

// ------------------------------------------------------
// -
// -    Fonctions
// -
// ------------------------------------------------------

async function all(params: GetAllParamsRaw): Promise<MaterialWithAvailability[]>;
async function all(params?: GetAllParamsPaginated): Promise<PaginatedData<MaterialWithAvailability[]>>;
async function all({ quantitiesPeriod, ...otherParams }: GetAllParamsPaginated | GetAllParamsRaw = {}): Promise<unknown> {
    const normalizedParams = {
        paginated: true,
        ...otherParams,
        ...quantitiesPeriod?.toQueryParams('quantitiesPeriod'),
    };

    const response = await requester.get('/materials', {
        params: normalizedParams,
    });

    const schema = normalizedParams.paginated
        ? withPaginationEnvelope(MaterialWithAvailabilitySchema)
        : MaterialWithAvailabilitySchema.array();

    return schema.parse(response.data);
}

const allWhileEvent = async (eventId: Event['id']): Promise<MaterialWithContext[]> => {
    const response = await requester.get(`/materials/while-event/${eventId}`);
    return MaterialWithContextSchema.array().parse(response.data);
};

const one = async (id: Material['id']): Promise<MaterialDetails> => {
    const response = await requester.get(`/materials/${id}`);
    return MaterialDetailsSchema.parse(response.data);
};

const create = async (data: MaterialEdit, onProgress?: ProgressCallback): Promise<MaterialDetails> => {
    const response = await requester.post('/materials', data, {
        ...(onProgress ? { onProgress } : {}),
    });
    return MaterialDetailsSchema.parse(response.data);
};

const update = async (id: Material['id'], data: Partial<MaterialEdit>, onProgress?: ProgressCallback): Promise<MaterialDetails> => {
    const response = await requester.put(`/materials/${id}`, data, {
        ...(onProgress ? { onProgress } : {}),
    });
    return MaterialDetailsSchema.parse(response.data);
};

const restore = async (id: Material['id']): Promise<MaterialDetails> => {
    const response = await requester.put(`/materials/${id}/restore`);
    return MaterialDetailsSchema.parse(response.data);
};

const remove = async (id: Material['id']): Promise<void> => {
    await requester.delete(`/materials/${id}`);
};

const bookings = async (id: Material['id'], params: GetBookingsParams = {}): Promise<PaginatedData<MaterialBookingSummary[]>> => {
    const normalizedParams = {
        ...omit(params, ['period']),
        ...params?.period?.toQueryParams('period'),
    };
    const response = await requester.get(`/materials/${id}/bookings`, { params: normalizedParams });
    return withPaginationEnvelope(MaterialBookingSummarySchema).parse(response.data);
};

const documents = async (id: Material['id']): Promise<Document[]> => {
    const response = await requester.get(`/materials/${id}/documents`);
    return DocumentSchema.array().parse(response.data);
};

const attachDocument = async (id: Material['id'], file: File, options: RequestConfig = {}): Promise<Document> => {
    const formData = new FormData(); formData.append('file', file);
    const response = await requester.post(`/materials/${id}/documents`, formData, options);
    return DocumentSchema.parse(response.data);
};

export default {
    all,
    allWhileEvent,
    one,
    create,
    update,
    restore,
    remove,
    bookings,
    documents,
    attachDocument,
};
