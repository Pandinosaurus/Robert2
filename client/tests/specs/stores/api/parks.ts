import requester from '@/globals/requester';
import apiParks from '@/stores/api/parks';
import materials from '@fixtures/materials';
import { withPaginationEnvelope } from '@fixtures/@utils';
import data from '@fixtures/parks';
import Decimal from 'decimal.js';

describe('Parks Api', () => {
    describe('all()', () => {
        it('parse the returned data correctly', async () => {
            const paginatedData = withPaginationEnvelope(data.default());
            jest.spyOn(requester, 'get').mockResolvedValue({ data: paginatedData });
            expect(await apiParks.all()).toMatchSnapshot();
        });
    });

    describe('list()', () => {
        it('parse the returned data correctly', async () => {
            jest.spyOn(requester, 'get').mockResolvedValue({ data: data.summary() });
            expect(await apiParks.list()).toMatchSnapshot();
        });
    });

    describe('one()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'get').mockResolvedValue({ data: datum });
            expect(await apiParks.one(datum.id)).toMatchSnapshot();
        });
    });

    describe('oneTotalAmount()', () => {
        it('parse the returned data correctly', async () => {
            jest.spyOn(requester, 'get').mockResolvedValue({ data: '119_061.80' });

            const result = await apiParks.oneTotalAmount(1);
            expect(result).toBeInstanceOf(Decimal);
            expect(result.toString()).toBe('119061.8');
        });
    });

    describe('materials()', () => {
        it('parse the returned data correctly', async () => {
            jest.spyOn(requester, 'get').mockResolvedValue({ data: materials.default() });
            expect(await apiParks.materials(1)).toMatchSnapshot();
        });
    });

    describe('create()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'post').mockResolvedValue({ data: datum });
            expect(await apiParks.create({} as any)).toMatchSnapshot();
        });
    });

    describe('update()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'put').mockResolvedValue({ data: datum });
            expect(await apiParks.update(datum.id, {} as any)).toMatchSnapshot();
        });
    });

    describe('restore()', () => {
        it.each(data.details())('parse the returned data correctly (with #$id)', async (datum: any) => {
            jest.spyOn(requester, 'put').mockResolvedValue({ data: datum });
            expect(await apiParks.restore(datum.id)).toMatchSnapshot();
        });
    });
});
