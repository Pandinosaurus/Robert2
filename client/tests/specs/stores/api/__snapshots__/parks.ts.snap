// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parks Api all() parse the returned data correctly 1`] = `
{
  "data": [
    {
      "country_id": 1,
      "id": 1,
      "locality": "Secretville",
      "name": "default",
      "note": null,
      "opening_hours": "Du lundi au vendredi, de 09:00 à 19:00.",
      "postal_code": "01234",
      "street": "Hangar 951",
      "total_items": 7,
      "total_stock_quantity": 87,
    },
    {
      "country_id": null,
      "id": 2,
      "locality": null,
      "name": "spare",
      "note": "Les bidouilles de fond de tiroir",
      "opening_hours": null,
      "postal_code": null,
      "street": null,
      "total_items": 2,
      "total_stock_quantity": 3,
    },
  ],
  "pagination": {
    "currentPage": 1,
    "perPage": 2,
    "total": {
      "items": 2,
      "pages": 1,
    },
  },
}
`;

exports[`Parks Api create() parse the returned data correctly (with #1) 1`] = `
{
  "country_id": 1,
  "has_ongoing_booking": false,
  "id": 1,
  "locality": "Secretville",
  "name": "default",
  "note": null,
  "opening_hours": "Du lundi au vendredi, de 09:00 à 19:00.",
  "postal_code": "01234",
  "street": "Hangar 951",
  "total_items": 7,
  "total_stock_quantity": 87,
}
`;

exports[`Parks Api create() parse the returned data correctly (with #2) 1`] = `
{
  "country_id": null,
  "has_ongoing_booking": false,
  "id": 2,
  "locality": null,
  "name": "spare",
  "note": "Les bidouilles de fond de tiroir",
  "opening_hours": null,
  "postal_code": null,
  "street": null,
  "total_items": 2,
  "total_stock_quantity": 3,
}
`;

exports[`Parks Api list() parse the returned data correctly 1`] = `
[
  {
    "id": 1,
    "name": "default",
  },
  {
    "id": 2,
    "name": "spare",
  },
]
`;

exports[`Parks Api materials() parse the returned data correctly 1`] = `
[
  {
    "attributes": [
      {
        "entities": [
          "material",
        ],
        "id": 1,
        "is_totalisable": true,
        "name": "Poids",
        "type": "float",
        "unit": "kg",
        "value": 36.5,
      },
      {
        "entities": [
          "material",
        ],
        "id": 2,
        "max_length": null,
        "name": "Couleur",
        "type": "string",
        "value": "Grise",
      },
      {
        "entities": [
          "material",
        ],
        "id": 3,
        "is_totalisable": true,
        "name": "Puissance",
        "type": "integer",
        "unit": "W",
        "value": 850,
      },
    ],
    "category_id": 1,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:00:00",
    "degressive_rate_id": 3,
    "description": "Console numérique 64 entrées / 8 sorties + Master + Sub",
    "id": 1,
    "is_discountable": false,
    "is_hidden_on_bill": false,
    "name": "Console Yamaha CL3",
    "note": null,
    "out_of_order_quantity": 1,
    "park_id": 1,
    "picture": "http://loxya.test/materials/1/picture",
    "reference": "CL3",
    "rental_price": /* Decimal Instance => */ "300",
    "replacement_price": /* Decimal Instance => */ "19400",
    "stock_quantity": 5,
    "sub_category_id": 1,
    "tags": [
      {
        "id": 1,
        "name": "Premium",
      },
    ],
    "tax_id": 1,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:00:00",
  },
  {
    "attributes": [
      {
        "entities": [
          "material",
        ],
        "id": 1,
        "is_totalisable": true,
        "name": "Poids",
        "type": "float",
        "unit": "kg",
        "value": 2.2,
      },
      {
        "entities": [
          "material",
        ],
        "id": 3,
        "is_totalisable": true,
        "name": "Puissance",
        "type": "integer",
        "unit": "W",
        "value": 35,
      },
    ],
    "category_id": 1,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:01:00",
    "degressive_rate_id": null,
    "description": "Système de diffusion numérique",
    "id": 2,
    "is_discountable": true,
    "is_hidden_on_bill": false,
    "name": "Processeur DBX PA2",
    "note": null,
    "out_of_order_quantity": 0,
    "park_id": 1,
    "picture": null,
    "reference": "DBXPA2",
    "rental_price": /* Decimal Instance => */ "25.5",
    "replacement_price": /* Decimal Instance => */ "349.9",
    "stock_quantity": 2,
    "sub_category_id": 2,
    "tags": [
      {
        "id": 1,
        "name": "Premium",
      },
    ],
    "tax_id": 1,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:02:00",
  },
  {
    "attributes": [
      {
        "entities": [
          "material",
        ],
        "id": 1,
        "is_totalisable": true,
        "name": "Poids",
        "type": "float",
        "unit": "kg",
        "value": 0.85,
      },
      {
        "entities": [
          "material",
        ],
        "id": 3,
        "is_totalisable": true,
        "name": "Puissance",
        "type": "integer",
        "unit": "W",
        "value": 150,
      },
    ],
    "category_id": 2,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:02:00",
    "degressive_rate_id": 1,
    "description": "Projecteur PAR64 à LED, avec son set de gélatines",
    "id": 3,
    "is_discountable": true,
    "is_hidden_on_bill": false,
    "name": "PAR64 LED",
    "note": "Soyez délicats avec ces projos !",
    "out_of_order_quantity": 4,
    "park_id": 1,
    "picture": null,
    "reference": "PAR64LED",
    "rental_price": /* Decimal Instance => */ "3.5",
    "replacement_price": /* Decimal Instance => */ "89",
    "stock_quantity": 34,
    "sub_category_id": 3,
    "tags": [
      {
        "id": 1,
        "name": "Premium",
      },
    ],
    "tax_id": null,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:02:00",
  },
  {
    "attributes": [
      {
        "entities": [
          "material",
        ],
        "id": 1,
        "is_totalisable": true,
        "name": "Poids",
        "type": "float",
        "unit": "kg",
        "value": 3.15,
      },
      {
        "entities": [
          "material",
        ],
        "id": 3,
        "is_totalisable": true,
        "name": "Puissance",
        "type": "integer",
        "unit": "W",
        "value": 60,
      },
      {
        "entities": [
          "material",
        ],
        "id": 4,
        "name": "Conforme",
        "type": "boolean",
        "value": true,
      },
    ],
    "category_id": 2,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:03:00",
    "degressive_rate_id": 1,
    "description": "Console DMX (jeu d'orgue) Showtec 6 canaux",
    "id": 4,
    "is_discountable": true,
    "is_hidden_on_bill": false,
    "name": "Showtec SDS-6",
    "note": null,
    "out_of_order_quantity": 0,
    "park_id": 1,
    "picture": null,
    "reference": "SDS-6-01",
    "rental_price": /* Decimal Instance => */ "15.95",
    "replacement_price": /* Decimal Instance => */ "59",
    "stock_quantity": 2,
    "sub_category_id": 4,
    "tags": [],
    "tax_id": 1,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:03:00",
  },
  {
    "attributes": [],
    "category_id": 1,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:14:00",
    "degressive_rate_id": 3,
    "description": "Câble audio XLR 10 mètres, mâle-femelle",
    "id": 5,
    "is_discountable": true,
    "is_hidden_on_bill": true,
    "name": "Câble XLR 10m",
    "note": null,
    "out_of_order_quantity": 8,
    "park_id": 1,
    "picture": null,
    "reference": "XLR10",
    "rental_price": /* Decimal Instance => */ "0.5",
    "replacement_price": /* Decimal Instance => */ "9.5",
    "stock_quantity": 40,
    "sub_category_id": null,
    "tags": [],
    "tax_id": 2,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:14:00",
  },
  {
    "attributes": [
      {
        "entities": [
          "material",
          "material",
        ],
        "id": 5,
        "name": "Date d'achat",
        "type": "date",
        "value": /* Day Instance => */ "2021-01-28",
      },
    ],
    "category_id": 1,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:15:00",
    "degressive_rate_id": 1,
    "description": "Mélangeur numérique 18 canaux",
    "id": 6,
    "is_discountable": false,
    "is_hidden_on_bill": false,
    "name": "Behringer X Air XR18",
    "note": null,
    "out_of_order_quantity": 1,
    "park_id": 1,
    "picture": null,
    "reference": "XR18",
    "rental_price": /* Decimal Instance => */ "49.99",
    "replacement_price": /* Decimal Instance => */ "419",
    "stock_quantity": 3,
    "sub_category_id": 1,
    "tags": [],
    "tax_id": 1,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:15:00",
  },
  {
    "attributes": [],
    "category_id": 3,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:16:00",
    "degressive_rate_id": 2,
    "description": "Volume utile: 9.3 m3",
    "id": 7,
    "is_discountable": false,
    "is_hidden_on_bill": false,
    "name": "Volkswagen Transporter",
    "note": null,
    "out_of_order_quantity": 0,
    "park_id": 2,
    "picture": null,
    "reference": "Transporter",
    "rental_price": /* Decimal Instance => */ "300",
    "replacement_price": /* Decimal Instance => */ "32000",
    "stock_quantity": 2,
    "sub_category_id": null,
    "tags": [],
    "tax_id": null,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:16:00",
  },
  {
    "attributes": [],
    "category_id": 4,
    "created_at": /* DateTime Instance => */ "2021-02-12 23:18:00",
    "degressive_rate_id": 3,
    "description": "Forêt mystique, typique des récits fantastiques.",
    "id": 8,
    "is_discountable": true,
    "is_hidden_on_bill": false,
    "name": "Décor Thème Forêt",
    "note": null,
    "out_of_order_quantity": 0,
    "park_id": 1,
    "picture": null,
    "reference": "Decor-Forest",
    "rental_price": /* Decimal Instance => */ "1500",
    "replacement_price": /* Decimal Instance => */ "8500",
    "stock_quantity": 2,
    "sub_category_id": null,
    "tags": [],
    "tax_id": 4,
    "updated_at": /* DateTime Instance => */ "2021-02-12 23:18:00",
  },
]
`;

exports[`Parks Api one() parse the returned data correctly (with #1) 1`] = `
{
  "country_id": 1,
  "has_ongoing_booking": false,
  "id": 1,
  "locality": "Secretville",
  "name": "default",
  "note": null,
  "opening_hours": "Du lundi au vendredi, de 09:00 à 19:00.",
  "postal_code": "01234",
  "street": "Hangar 951",
  "total_items": 7,
  "total_stock_quantity": 87,
}
`;

exports[`Parks Api one() parse the returned data correctly (with #2) 1`] = `
{
  "country_id": null,
  "has_ongoing_booking": false,
  "id": 2,
  "locality": null,
  "name": "spare",
  "note": "Les bidouilles de fond de tiroir",
  "opening_hours": null,
  "postal_code": null,
  "street": null,
  "total_items": 2,
  "total_stock_quantity": 3,
}
`;

exports[`Parks Api restore() parse the returned data correctly (with #1) 1`] = `
{
  "country_id": 1,
  "has_ongoing_booking": false,
  "id": 1,
  "locality": "Secretville",
  "name": "default",
  "note": null,
  "opening_hours": "Du lundi au vendredi, de 09:00 à 19:00.",
  "postal_code": "01234",
  "street": "Hangar 951",
  "total_items": 7,
  "total_stock_quantity": 87,
}
`;

exports[`Parks Api restore() parse the returned data correctly (with #2) 1`] = `
{
  "country_id": null,
  "has_ongoing_booking": false,
  "id": 2,
  "locality": null,
  "name": "spare",
  "note": "Les bidouilles de fond de tiroir",
  "opening_hours": null,
  "postal_code": null,
  "street": null,
  "total_items": 2,
  "total_stock_quantity": 3,
}
`;

exports[`Parks Api update() parse the returned data correctly (with #1) 1`] = `
{
  "country_id": 1,
  "has_ongoing_booking": false,
  "id": 1,
  "locality": "Secretville",
  "name": "default",
  "note": null,
  "opening_hours": "Du lundi au vendredi, de 09:00 à 19:00.",
  "postal_code": "01234",
  "street": "Hangar 951",
  "total_items": 7,
  "total_stock_quantity": 87,
}
`;

exports[`Parks Api update() parse the returned data correctly (with #2) 1`] = `
{
  "country_id": null,
  "has_ongoing_booking": false,
  "id": 2,
  "locality": null,
  "name": "spare",
  "note": "Les bidouilles de fond de tiroir",
  "opening_hours": null,
  "postal_code": null,
  "street": null,
  "total_items": 2,
  "total_stock_quantity": 3,
}
`;
